crates/typst/src/geom/color.rs

```diff
@@ -334,183 +954,696 @@ cast! {
 }
 
 /// A weight for color mixing.
-struct Weight(f32);
+struct Weight(f64);
 
 cast! {
     Weight,
-    v: f64 => Self(v as f32),
-    v: Ratio => Self(v.get() as f32),
+    v: f64 => Self(v),
+    v: Ratio => Self(v.get()),
 }
 
 /// Convert an RGBA color to four components in the given color space.
-fn rgba_to_vec4(color: RgbaColor, space: ColorSpace) -> [f32; 4] {
+pub fn color_to_vec4(color: Color, space: ColorSpace) -> [f64; 4] {
     match space {
-        ColorSpace::Oklab => {
-            let RgbaColor { r, g, b, a } = color;
-            let oklab = oklab::srgb_to_oklab(oklab::RGB { r, g, b });
-            [oklab.l, oklab.a, oklab.b, a as f32 / 255.0]
-        }
-        ColorSpace::Srgb => {
-            let RgbaColor { r, g, b, a } = color;
-            [r as f32 / 255.0, g as f32 / 255.0, b as f32 / 255.0, a as f32 / 255.0]
-        }
+        ColorSpace::Oklab => color.to_oklab().to_vec4(),
+        ColorSpace::Srgb => color.to_rgba().to_vec4(),
+        ColorSpace::LinearRGB => color.to_linear_rgb().to_vec4(),
+        ColorSpace::Hsl => color.to_hsl().to_vec4(),
+        ColorSpace::Hsv => color.to_hsv().to_vec4(),
+        ColorSpace::Cmyk => color.to_cmyk().to_vec4(),
+        ColorSpace::D65Gray => color.to_luma().to_vec4(),
     }
 }
 
 /// Convert four components in the given color space to RGBA.
-fn vec4_to_rgba(vec: [f32; 4], space: ColorSpace) -> RgbaColor {
+fn vec4_to_color(vec: [f64; 4], space: ColorSpace) -> Color {
     match space {
-        ColorSpace::Oklab => {
-            let [l, a, b, alpha] = vec;
-            let oklab::RGB { r, g, b } = oklab::oklab_to_srgb(oklab::Oklab { l, a, b });
-            RgbaColor { r, g, b, a: (alpha * 255.0).round() as u8 }
-        }
-        ColorSpace::Srgb => {
-            let [r, g, b, a] = vec;
-            RgbaColor {
-                r: (r * 255.0).round() as u8,
-                g: (g * 255.0).round() as u8,
-                b: (b * 255.0).round() as u8,
-                a: (a * 255.0).round() as u8,
+        ColorSpace::Oklab => Color::Oklab(OklabColor::from_vec4(vec)),
+        ColorSpace::Srgb => Color::Rgba(RgbaColor::from_vec4(vec)),
+        ColorSpace::LinearRGB => Color::LinearRgb(LinearRgbColor::from_vec4(vec)),
+        ColorSpace::Hsl => Color::Hsl(HslColor::from_vec4(vec)),
+        ColorSpace::Hsv => Color::Hsv(HsvColor::from_vec4(vec)),
+        ColorSpace::Cmyk => Color::Cmyk(CmykColor::from_vec4(vec)),
+        ColorSpace::D65Gray => Color::Luma(LumaColor::from_vec4(vec)),
+    }
+}
+
+/// A color space for mixing.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Cast)]
+pub enum ColorSpace {
```

Comment:
Should this really cast from strings? Or do we want to use the constructors as the kinds? as with `color.kind()` (I guess it could be named `color.space()` ...)

Or is this something else entirely which just happens to have quite similar variants?

Reply:
They used to be different (in the initial implementation of color mixing) but it's true that it doesn't make much sense now, I'll change it to work from the color function, but internally it has to stay, it just won't be user facing anymore!